xds/src/main/java/com/linecorp/centraldogma/xds/internal/ControlPlanePlugin.java

```diff
@@ -0,0 +1,343 @@
+                                    repository, fileName, watchedRevision, snapshotConverter));
+                        } else {
+                            CONTROL_PLANE_EXECUTOR.schedule(
+                                    () -> watchRepository(
+                                            repository, fileName, watchedRevision, snapshotConverter),
+                                    BACKOFF_SECONDS, TimeUnit.SECONDS);
+                        }
+                        return null;
+                    },
+                    CONTROL_PLANE_EXECUTOR);
+            return null;
+        }, CONTROL_PLANE_EXECUTOR);
+    }
+
+    private boolean clusters(Collection<Entry<?>> entries, Revision revision, SimpleCache<String> cache) {
+        final Builder<Cluster> clustersBuilder = ImmutableList.builder();
+        for (Entry<?> entry : entries) {
+            try {
+                final Cluster.Builder clusterBuilder = Cluster.newBuilder();
+                JsonFormatUtil.parser().merge(entry.contentAsText(), clusterBuilder);
+                clustersBuilder.add(clusterBuilder.build());
+            } catch (Throwable e) {
+                logger.warn("Unexpected exception is raised while building a cluster using {}" +
+                            ". Try watching after {} seconds..", entry, BACKOFF_SECONDS, e);
+                return false;
+            }
+        }
+
+        setNewSnapshot(cache, ResourceType.CLUSTER,
+                       CentralDogmaSnapshotResources.create(clustersBuilder.build(), revision));
+        return true;
+    }
+
+    private boolean endpoints(Collection<Entry<?>> entries, Revision revision, SimpleCache<String> cache) {
+        final Builder<ClusterLoadAssignment> endpointsBuilder = ImmutableList.builder();
+        for (Entry<?> entry : entries) {
+            try {
+                final ClusterLoadAssignment.Builder endpointBuilder = ClusterLoadAssignment.newBuilder();
+                JsonFormatUtil.parser().merge(entry.contentAsText(), endpointBuilder);
+                endpointsBuilder.add(endpointBuilder.build());
+            } catch (Throwable e) {
+                logger.warn("Unexpected exception is raised while building an endpoint using {}" +
+                            ". Try watching after {} seconds..", entry, BACKOFF_SECONDS, e);
+                return false;
+            }
+        }
+
+        setNewSnapshot(cache, ResourceType.ENDPOINT,
+                       CentralDogmaSnapshotResources.create(endpointsBuilder.build(), revision));
+        return true;
+    }
+
+    private boolean listeners(Collection<Entry<?>> entries, Revision revision, SimpleCache<String> cache) {
+        final Builder<Listener> listenersBuilder = ImmutableList.builder();
+        for (Entry<?> entry : entries) {
+            try {
+                final Listener.Builder listenerBuilder = Listener.newBuilder();
+                JsonFormatUtil.parser().merge(entry.contentAsText(), listenerBuilder);
```

Comment:
Optional) We may consider using [protobuf-jackson](https://github.com/curioswitch/protobuf-jackson) for better performance some time later.

Reply:
I completely forgot about it. 😓 Let me handle it in this PR. Thanks!

Reply:
Let's do that later since it requires registering all messages.
Will handle this later when I automate the process. 😉